spec/controllers/omniauth_callbacks_controller_spec.rb

```diff
@@ -115,5 +119,89 @@
       end
     end
   end
-end
 
+  describe 'GET #facebook' do
+    before { stub_env_for_omniauth_fb }
+    let(:oauth_email) { request.env['omniauth.auth']['info']['email'] }
+    let(:oauth_user) { User.find_by(email: oauth_email) }
+
+    context 'when facebook email doesnt exist in the system' do
+      let(:user) { oauth_user }
+
+      before { get :facebook }
+
+      it 'creates user with info in facebook' do
+        expect(user.name).to eq 'Test User'
+      end
+
+      include_examples 'successful sign in with oauth details'
+    end
+
+    context 'when facebook email already exist in the system' do
+      let!(:user) { create(:user, email: 'example@xyze.it') }
+
+      before { get :facebook }
+
+      it 'updates the user with facebook credentials' do
+        expect(user.reload.token).to eq 'abcdefg12345'
+      end
+
+      include_examples 'successful sign in with oauth details'
+    end
+
+    context 'when facebook enabling fails' do
+      before do
+        stub_env_for_omniauth_fb
+        allow(User).to receive(:find_for_oauth)
+          .with(request.env['omniauth.auth']).and_return(nil)
+        get :facebook
+      end
+
+      it 'redirects to sign in path' do
+        expect(response).to redirect_to(new_user_session_path)
+      end
+    end
+
+    context 'when an invitation_token is passed in' do
+      let(:inviter) { create(:user) }
+      let(:invitee_email) { oauth_email }
+      let!(:invitee) { User.invite!({ email: invitee_email }, inviter) }
+
+      before do
+        request.env['omniauth.params'] = { 'invitation_token' => invitee.invitation_token }
+        get :facebook
+      end
+
+      context 'when the user logs in with the same email as the invitation' do
+        let(:user) { invitee }
+
+        include_examples 'successful sign in with oauth details'
+        it { expect(user.reload.invitation_accepted_at).to be_present }
+      end
+
+      context 'when the user logs in with a different email from the invitation' do
+        let(:invitee_email) { 'no-invite@xyze.it' }
+        let(:user) { oauth_user }
+
+        include_examples 'successful sign in with oauth details'
+
+        it { expect(user.reload.invitation_accepted_at).to be_blank }
+      end
+    end
+
+    context 'user avatar image uploads' do
+      before { stub_env_for_omniauth_fb }
+      let(:oauth_email) { request.env['omniauth.auth']['info']['email'] }
+      let(:oauth_user) { User.find_by(email: oauth_email) }
+      let(:user) { oauth_user }
+      before { get :facebook }
+
+      context 'when third party avatar is nil' do
```

Comment:
What happens when the third party avatar is not nil? I think we should test for that case as well :)